katharsis-core/src/main/java/io/katharsis/resource/registry/responseRepository/ResponseRepository.java

```diff
@@ -33,66 +45,218 @@
+        this.moduleRegistry = moduleRegistry;
         this.resourceInformation = resourceInformation;
     }
 
-    @SuppressWarnings("rawtypes")
-	protected JsonApiResponse getResponse(Object repository, Object resource, RequestSpec requestSpec) {
-        if (resource instanceof JsonApiResponse) {
-            return (JsonApiResponse) resource;
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+	protected <T> Iterable<T> filterResult(Iterable<?> resources, RepositoryRequestSpec requestSpec) {
+    	RepositoryResultFilterChainImpl<T> chain = new RepositoryResultFilterChainImpl<>((Iterable)resources);
+    	return chain.doFilter(newRepositoryFilterContext(requestSpec));
+    }
+    
+	protected JsonApiResponse getResponse(Object repository, Object result, RepositoryRequestSpec requestSpec) {
+        if (result instanceof JsonApiResponse) {
+            return (JsonApiResponse) result;
         }
 
-        Iterable resources;
-        if (resource instanceof Iterable) {
-            resources = (Iterable) resource;
+        Iterable<?> resources;
+        boolean isCollection = result instanceof Iterable;
+        if (isCollection) {
+            resources = (Iterable<?>) result;
         } else {
-            resources = Collections.singletonList(resource);
+            resources = Collections.singletonList(result);
         }
+        Iterable<?> filteredResult = filterResult(resources, requestSpec);
         MetaInformation metaInformation = getMetaInformation(repository, resources, requestSpec);
         LinksInformation linksInformation = getLinksInformation(repository, resources, requestSpec);
 
-        return new JsonApiResponse().setEntity(resource).setLinksInformation(linksInformation)
+        Object resultEntity;
+        if(isCollection){
+        	resultEntity = filteredResult;
+        }else {
+        	Iterator<?> iterator = filteredResult.iterator();
+        	if(iterator.hasNext()){
+        		resultEntity = iterator.next();
+        		PreconditionUtil.assertFalse("expected unique result", iterator.hasNext());
+        	}else{
+        		resultEntity = null;
+        	}
+        }
+        
+        return new JsonApiResponse().setEntity(resultEntity).setLinksInformation(linksInformation)
                 .setMetaInformation(metaInformation);
     }
 
+    private MetaInformation getMetaInformation(Object repository, Iterable<?> resources, RepositoryRequestSpec requestSpec) {
+    	RepositoryMetaFilterChainImpl chain = new RepositoryMetaFilterChainImpl(repository);
+    	return chain.doFilter(newRepositoryFilterContext(requestSpec), resources);
+    }
+    
     @SuppressWarnings({ "unchecked", "rawtypes" })
-    private MetaInformation getMetaInformation(Object repository, Iterable<?> resources, RequestSpec requestSpec) {
+    private MetaInformation doGetMetaInformation(Object repository, Iterable<?> resources, RepositoryRequestSpec requestSpec) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Move this method into "RepositoryMetaFilterChainImpl". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS3398)